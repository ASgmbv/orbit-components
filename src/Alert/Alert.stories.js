// @flow
import * as React from "react";
import { storiesOf, setAddon } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import styles from "@sambego/storybook-styles";
import chaptersAddon from "react-storybook-addon-chapters";
import { withKnobs, text, boolean, select } from "@storybook/addon-knobs/react";

import * as Icons from "../icons";

import Alert from "./index";

setAddon(chaptersAddon);

const options = {
  showSource: true,
  allowSourceToggling: false,
  showPropTables: false,
  allowPropTablesToggling: false,
};

storiesOf("Alert", module)
  .addDecorator(withKnobs)
  .addDecorator(
    styles({
      padding: "20px",
    }),
  )
  .addWithChapters("Default alert", () => {
    const message = text("Message", "The quick, brown fox jumps over a lazy dog.");
    return {
      info: "Some description about this type of alert in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Alert>{message}</Alert>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Without icon", () => {
    const message = text("Message", "The quick, brown fox jumps over a lazy dog.");
    return {
      info: "Some description about this type of alert in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Alert noIcon>{message}</Alert>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("With title", () => {
    const title = text("Title", "Alert with title");
    const message = text("Message", "The quick, brown fox jumps over a lazy dog.");
    return {
      info: "Some description about this type of alert in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Alert title={title}>{message}</Alert>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("With title and without icon", () => {
    const title = text("Title", "Alert with title & without icon");
    const message = text("Message", "The quick, brown fox jumps over a lazy dog.");
    return {
      info: "Some description about this type of alert in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <Alert title={title} noIcon>
                  {message}
                </Alert>
              ),
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Playground", () => {
    const typeOptions = {
      info: "info",
      success: "success",
      warning: "warning",
      critical: "critical",
    };
    const type = select("Type", typeOptions, "info");
    const title = text("Title", "Customisable alert title");
    const message = text("Message", "Message is customisable too");
    const closable = boolean("Closable", false);
    const noIcon = boolean("No icon", false);
    const source = select("Icon", Object.keys(Icons));
    return {
      info: "Some description about this type of alert in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <Alert
                  type={type}
                  Icon={Icons[source]}
                  title={title}
                  closable={closable}
                  noIcon={noIcon}
                  onClose={action("Close")}
                >
                  {message}
                </Alert>
              ),
              options,
            },
          ],
        },
      ],
    };
  });
