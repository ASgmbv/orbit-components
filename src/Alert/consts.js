// @flow
export const TYPE_OPTIONS = {
  INFO: "info",
  SUCCESS: "success",
  WARNING: "warning",
  CRITICAL: "critical",
};

export const TOKENS = {
  backgroundAlert: "backgroundAlert",
  colorIconAlert: "colorIconAlert",
  colorTextAlert: "colorTextAlert",
};
