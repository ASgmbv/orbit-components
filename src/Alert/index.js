// @flow
import * as React from "react";
import * as tokens from "@kiwicom/orbit-design-tokens";

import * as Icons from "../icons";
import { Button } from "../index";

const Icon = ({ type, Icon: CustomIcon, customColor }) => {
  switch (type) {
    case "info":
      return (
        (CustomIcon && <CustomIcon customColor={customColor} />) || (
          <Icons.InformationCircle customColor={customColor} />
        )
      );
    case "success":
      return (
        (CustomIcon && <CustomIcon customColor={customColor} />) || (
          <Icons.Check customColor={customColor} />
        )
      );
    case "warning":
      return (
        (CustomIcon && <CustomIcon customColor={customColor} />) || (
          <Icons.Alert customColor={customColor} />
        )
      );
    case "critical":
      return (
        (CustomIcon && <CustomIcon customColor={customColor} />) || (
          <Icons.AlertCircle customColor={customColor} />
        )
      );
    default:
      return null;
  }
};

const backgroundAlert = {
  info: tokens.backgroundAlertInfo,
  success: tokens.backgroundAlertSuccess,
  warning: tokens.backgroundAlertWarning,
  critical: tokens.backgroundAlertCritical,
};

const colorTextAlert = {
  info: tokens.colorTextAlertInfo,
  success: tokens.colorTextAlertSuccess,
  warning: tokens.colorTextAlertWarning,
  critical: tokens.colorTextAlertCritical,
};

const colorIconAlert = {
  info: tokens.colorAlertIconInfo,
  success: tokens.colorAlertIconSuccess,
  warning: tokens.colorAlertIconWarning,
  critical: tokens.colorAlertIconCritical,
};

type Props = {
  type: $Keys<typeof backgroundAlert | typeof colorTextAlert>,
  title?: string,
  Icon?: React.ComponentType<*>,
  closable: boolean,
  noIcon: boolean,
  onClose: () => void,
  children: React.Node,
};

const Alert = (props: Props) => {
  const { type, title, closable, noIcon, onClose, children } = props;
  return (
    <div className="alert">
      <div className="alertIcon">
        {!noIcon &&
          (Icon && <Icon type={type} Icon={props.Icon} customColor={colorIconAlert[type]} />)}
      </div>
      <div>
        {title && <div className="alertTitle">{title}</div>}
        <p className="alertContent">{children}</p>
      </div>
      {closable && (
        <div className="alertClose">
          <Button
            onClick={onClose}
            type={type}
            variation="link"
            size="small"
            Icon={Icons.CloseCircle}
            onlyIcon
          />
        </div>
      )}
      <style jsx>
        {`
          .alert {
            position: relative;
            display: flex;
            width: 100%;
            padding: ${!noIcon
              ? `${tokens.paddingAlert} ${tokens.paddingAlertWithIcon}`
              : `${tokens.paddingAlert}`};
            padding-right: ${closable ? tokens.spaceXXLarge : "auto"};
            border-radius: ${tokens.borderRadiusNormal};
            background: ${backgroundAlert[type]};
            color: ${colorTextAlert[type]};
            font-family: ${tokens.fontFamily};
            font-size: ${tokens.fontSizeTextNormal};
            box-sizing: border-box;
          }
          .alertInner {
            display: flex;
          }
          .alertIcon {
            flex-shrink: 0;
            margin-right: ${tokens.spaceSmall};
          }
          .alertTitle {
            display: flex;
            align-items: center;
            margin-bottom: ${tokens.spaceXSmall};
            font-weight: ${tokens.fontWeightBold};
            line-height: ${tokens.lineHeightHeading};
            min-height: ${tokens.heightIconMedium};
          }
          .alertContent {
            display: flex;
            align-items: center;
            min-height: ${!title ? tokens.heightIconMedium : `0`};
            margin: ${title ? `0 0 ${tokens.spaceXXSmall} 0` : `0`};
          }
          .alertClose {
            position: absolute;
            top: ${tokens.spaceXXSmall};
            right: ${tokens.spaceXXSmall};
          }
        `}
      </style>
    </div>
  );
};

Alert.defaultProps = {
  type: "info",
  title: "",
  closable: false,
  noIcon: false,
  onClose: () => {},
};

export default Alert;
