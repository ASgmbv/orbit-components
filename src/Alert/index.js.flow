// @flow
import defaultTokens from "../defaultTokens";

type Type = "info" | "success" | "warning" | "critical";

export type Props = {|
  +type?: Type,
  +children?: React$Node,
  +title?: string | React$Node,
  +icon?: React$Element<any> | boolean,
  +closable?: boolean,
  +onClose?: () => void,
  +theme?: typeof defaultTokens,
|};

declare export default React$ComponentType<Props>;
