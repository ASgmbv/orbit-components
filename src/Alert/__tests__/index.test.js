// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Alert from "../index";

describe("Rendered with children", () => {
  it("Should contain message", () => {
    const message = "Alert message";
    const component = shallow(<Alert>{message}</Alert>);
    expect(component).toMatchSnapshot();
  });
});

describe("Called when the closable button is clicked", () => {
  it("Should call onClose function", () => {
    const onClose = jest.fn();
    const message = "Alert message";
    const component = shallow(
      <Alert onClose={onClose} closable>
        {message}
      </Alert>,
    );
    component
      .find(".alertClose")
      .children()
      .simulate("click");
    expect(onClose).toHaveBeenCalled();
  });
});
