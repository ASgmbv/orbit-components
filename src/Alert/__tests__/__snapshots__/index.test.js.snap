// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered with children Should contain message 1`] = `
<div
  className="jsx-2358314486 alert"
>
  <div
    className="jsx-2358314486 alertIcon"
  >
    <Icon
      customColor="#0176D2"
      type="info"
    />
  </div>
  <div
    className="jsx-2358314486"
  >
    <p
      className="jsx-2358314486 alertContent"
    >
      Alert message
    </p>
  </div>
  <JSXStyle
    css=".alert.__jsx-style-dynamic-selector{position:relative;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;width:100%;padding:16px 12px;padding-right:auto;border-radius:3px;background:#e0f6ff;color:#07405c;font-family:\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif;font-size:14px;box-sizing:border-box;}.alertInner.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.alertIcon.__jsx-style-dynamic-selector{-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;margin-right:12px;}.alertTitle.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;margin-bottom:8px;font-weight:700;line-height:1.2;min-height:24px;}.alertContent.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;min-height:24px;margin:0;}.alertClose.__jsx-style-dynamic-selector{position:absolute;top:4px;right:4px;}"
    dynamic={
      Array [
        "16px 12px",
        "auto",
        "3px",
        "#e0f6ff",
        "#07405c",
        "\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif",
        "14px",
        "12px",
        "8px",
        "700",
        "1.2",
        "24px",
        "24px",
        "0",
        "4px",
        "4px",
      ]
    }
    styleId="1379871782"
  />
</div>
`;
